src/planner/subquery/rewrite_correlated_expressions.cpp

```diff
@@ -5,34 +5,80 @@
 #include "duckdb/planner/expression/bound_constant_expression.hpp"
 #include "duckdb/planner/expression/bound_operator_expression.hpp"
 #include "duckdb/planner/expression/bound_subquery_expression.hpp"
+#include "duckdb/planner/query_node/bound_select_node.hpp"
 #include "duckdb/planner/expression_iterator.hpp"
+#include "duckdb/planner/tableref/bound_joinref.hpp"
+#include "duckdb/planner/operator/logical_dependent_join.hpp"
 
 namespace duckdb {
 
 RewriteCorrelatedExpressions::RewriteCorrelatedExpressions(ColumnBinding base_binding,
-                                                           column_binding_map_t<idx_t> &correlated_map)
-    : base_binding(base_binding), correlated_map(correlated_map) {
+                                                           column_binding_map_t<idx_t> &correlated_map,
+                                                           idx_t lateral_depth, bool recursive_rewrite)
+    : base_binding(base_binding), correlated_map(correlated_map), lateral_depth(lateral_depth),
+      recursive_rewrite(recursive_rewrite) {
 }
 
 void RewriteCorrelatedExpressions::VisitOperator(LogicalOperator &op) {
+	if (recursive_rewrite) {
+		// recursively visit all the nodes and update column bindings
+		switch (op.type) {
+		case LogicalOperatorType::LOGICAL_DEPENDENT_JOIN:
+			// Need to specially handle depths for dependendent joins
+			D_ASSERT(op.children.size() == 2);
+
+			if (op.swapped_children) {
+				lateral_depth++;
+				VisitOperator(*op.children[0]);
+				lateral_depth--;
+				VisitOperator(*op.children[1]);
+			} else {
+				VisitOperator(*op.children[0]);
+				// Increment the lateral depth so that relevant correlated columns can be identified correctly
+				lateral_depth++;
+				VisitOperator(*op.children[1]);
+				// Reset depth to original state
+				lateral_depth--;
+			}
+			break;
+		default:
+			VisitOperatorChildren(op);
+		}
+	}
+	// update the bindings in the correlated columns of the dependendent join
+	if (op.type == LogicalOperatorType::LOGICAL_DEPENDENT_JOIN) {
+		auto &plan = (LogicalDependentJoin &)op;
```

Comment:
Could we use the new `Cast` syntax here (`op.Cast< LogicalDependentJoin>()`)?

Reply:
Sure, we'll make that change!